### PR TITLE
vfs: cleanup on max_open_files threshold

### PR DESCRIPTION
**Pull request by barraguda**
## Problem

https://github.com/kinode-dao/kinode/issues/525

## Solution

If open_files.len() is above threshold (180 right now based on my mac testing), run a cleanup that prunes 50% of the open files. 

Interval cleanup of idle files still happens, but now in a tokio::select loop instead.

## Notes

Linux seems to have a larger supported open filehandles amount per OS process (~1000 as opposed to mac's 256)

### CODE REVIEW

File: kinode/src/vfs.rs [hunk 1]
```diff
@@ -594,26 +593,49 @@ async fn open_file<P: AsRef<Path>>(
...
+        .iter()
+        .map(|entry| (entry.key().clone(), entry.value().1))
+        .collect();
+    entries.sort_by(|a, b| a.1.cmp(&b.1));
+
+    // remove oldest entries until we reach the target count
+    for (path, _) in entries.into_iter().take(current_count - target_count) {
+        // use remove_if_present to avoid potential race conditions
+        open_files.remove_if(&path, |_, _| true);
+    }
```
Comment (dr-frmr) [inline kinode/src/vfs.rs]: is there a way to make this more efficient using `retain` on the dashmap with a fixed duration calculated by the average of the map? could be better than cloning every path. something like

1. calculate average last access time by traversing map once
2. retain entries only with a below average last access time

Comment (barraguda) [reply kinode/src/vfs.rs]: this is a good point, I'll try something out!

**Comment by barraguda** [comment]
some notes for the other suggestions during discussions: 

1. matching upon a specific "TooManyOpenFiles" error in the open_file() function seems to me slightly too risky, there's no specific error kind like that in tokio::fs and matching on raw OS error codes that may be differ in environments seems risky
2. closing 1 random file every time we reach the limit rather than 50% of them has around the same performance for 5000 sequential file opens (~700ms-900ms on my machine, debug binary)

**Comment by nick1udwig** [comment]
1. Surprised about this. I agree matching on raw OS error codes seems less good
2. Gotta test performance on release; debug is meaningless

**Comment by dr-frmr** [comment]
Possible/reasonable to match on *any* error coming out of open_file()?

**Comment by barraguda** [comment]
> Possible/reasonable to match on _any_ error coming out of open_file()?

I like this idea, but noting that keeping the open_files very close to the limit and relying on specific errors like this might be brittle. 

Another big note is, we need to save cursor positions of files we close, so that processes that expect them to be at a certain location won't notice a difference even if the vfs closed and reopened them. This is ok but leads to us not being able to use synchronous functions like dashmap.retain() and makes closing a lot of files considerable slower..

**Review by dr-frmr — APPROVED** [review]
Nice refactor, looks clean

**Pull request merged**